src/Compilers/Core/Portable/InternalUtilities/SemaphoreSlimExtensions.cs

```diff
@@ -35,7 +35,23 @@ public SemaphoreDisposer(SemaphoreSlim semaphore)
 
             public void Dispose()
             {
-                _semaphore.Release();
+                // Officially, Dispose() being called more than once is allowable, but in this case
+                // if that were to ever happen that means something is very, very wrong. Since it's an internal
+                // type, better to be strict.
+
+                // Nulling this out also means it's a bit easier to diagnose some async deadlocks; if you have an
+                // async deadlock where a SemaphoreSlim is held but you're unsure why, as long all the users of the
+                // SemaphoreSlim used the Disposable helpers, you can search memory and find the instance that
+                // is pointing to the SemaphoreSlim that hasn't nulled out this field yet; in that case you know
+                // that's holding the lock and can figure out who is holding that SemaphoreDisposer.
```

Comment:
we're in a struct.  so this does only work as long as the NonCopyable part above really is maintained.  i can allow this :)

Reply:
@CyrusNajmabadi Indeed, the [NonCopyable] being violated does mean we could end up with "surprises", but at this point, this is the least of your concerns IMHO. :smile:

Reply:
> so this does only work as long as the NonCopyable part above really is maintained

The type is copied twice in this very file so `[NonCopyable]` is more like `[NonCopyableForSomeVersionsOfCopy]`

Reply:
> Instances of this type are copied twice in this very file

There are no copies of this type in this file (moves are not copies).

Reply:
There are two returns which are copies. You label them moves but that doesn't change that copies occur.